nearcache/config.go

```diff
@@ -0,0 +1,250 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package nearcache
+
+import (
+	"math"
+
+	"github.com/hazelcast/hazelcast-go-client/internal/check"
+	ihzerrors "github.com/hazelcast/hazelcast-go-client/internal/hzerrors"
+)
+
+const (
+	defaultMaxEntryCount            = 10_000
+	defaultEvictionPolicy           = EvictionPolicyLRU
+	defaultStoreInitialDelaySeconds = 600
+	defaultStoreIntervalSeconds     = 600
+	defaultMemoryFormat             = InMemoryFormatBinary
+)
+
+// Config is the Near Cache configuration.
+type Config struct {
+	invalidateOnChange *bool
+	// Name is the name of this Near Cache configuration.
+	Name string
+	// EvictionConfig is the optional eviction configuration for the Near Cache.
+	EvictionConfig EvictionConfig
+	// PreloaderConfig is the optional preloader configuration for the Near Cache.
+	PreloaderConfig PreloaderConfig
+	// InMemoryFormat specifies how the entry values are stored in the Near Cache.
+	// InMemoryFormatBinary stores the values after serializing them.
+	// InMemoryFormatObject stores the values in their original form.
+	// The default is InMemoryFormatBinary.
+	InMemoryFormat InMemoryFormat
+	// SerializeKeys specifies how the entry keys are stored in the Near Cache.
+	// If false, keys are stored in their original form.
+	// If true, keys are stored after serializing them.
+	// Storing keys in serialized form is required when the key is not hashable, such as slices.
+	// The default is false.
+	SerializeKeys bool
+	// TimeToLiveSeconds is the maximum number of seconds for each entry to stay in the Near Cache (time to live).
+	// Entries that are older than TimeToLiveSeconds will automatically be evicted from the Near Cache.
+	// Must be non-negative.
+	// The value 0 equals to math.MaxInt32
+	// The default is 0.
+	TimeToLiveSeconds int
+	//MaxIdleSeconds is the maximum number of seconds each entry can stay in the Near Cache as untouched (not-read).
+	//Entries that are not read (touched) more than MaxIdleSeconds value will get removed from the Near Cache.
+	//Accepts any integer between {@code 0} and {@link Integer#MAX_VALUE}.
+	// Must be non-negative.
+	// The value 0 equals to math.MaxInt32
+	// The default is 0.
+	MaxIdleSeconds int
+}
+
+// Clone returns a copy of the configuration.
+func (c Config) Clone() Config {
+	return Config{
+		invalidateOnChange: c.invalidateOnChange,
```

Comment:
That's not necessary, since `SetInvalidateOnChange` sets a new reference to `invalidateOnChange` every time it is called. Here's a test:
```go
ncc := nearcache.Config{
    Name:           mapName,
}
ncc.SetInvalidateOnChange(false)
ncc2 := ncc.Clone()
fmt.Println(ncc2.InvalidateOnChange())
// prints false
ncc.SetInvalidateOnChange(true)
fmt.Println(ncc2.InvalidateOnChange())
// prints false
```